### Tittle: Telco Churn Classification Project

In this project,we aim to find the likelihood of a customer leaving an organization, the key indicators of churn, as well as the retention strategies that can be implemented to avert this problem, thus, we focus on building classification models to perform churn analysis on customer data, a critical task for companies looking to enhance their revenue by retaining customers.

## 1.0 Business Understanding

### 1.1 Introduction
Customer churn  is a significant problem in the telecom industry as it results in reduced profit margin and negatively impacting long-term sustainability. 
Churn, which refers to customers discontinuing their service and moving to a competitor, can be driven by various factors such as charges, customer service quality, network coverage, and the competitiveness of offerings. The implications of high churn rates are:

* Revenue Loss
* Decreased ROI on marketing
* Reputational Damage due to Customer dissatisfaction
* Reduced Market Share and Growth
* Lower Employee Morale
* Financial Uncertainty

Due to this, Machne Learning and Advanced Analytics has proided us with the technologies to transform raw data into actions and insights. We will employ Classification models to get actionable insights.

Classification in Machine Learning is a supervised learning approach where the computer program learns from provided data to make new observations by classifying. Various classification algorithms such as logistic regression, decision trees, random forests, and gradient boosting will be explored to identify the most effective model for the given dataset.
The **objective** is to determine the class or category into which new data points will fall. 

In this project, an elaborate analysis will be conducted to train at least some models for predicting customer churn in a telecom company. This analysis will adhere to the **CRISP-DM framework**, ensuring a structured and systematic approach to model development and evaluation.

### 1.2 Project Objective
Objective of this project is to develop a classification model for churn analysis which is to predict whether customers are likely to leave or continue their relationship with the company. By identifying customers at risk of churning, the company can take proactive measures to retain them, thus increasing revenue and profit margins.

### 1.3 Data Description
The project will utilize historical data that contains details on customer behaviours and transactional details.
Datasets will be retrieved from various sources including a database, GitHub repository and Onedrive.


### 1.4 Success metrics
- `Good:` accurately predicting churn at least 75% measured with the harmonic f1-score metric. 

- `Excellent:` accurately predicting churn at least 80%.


### 1.5 Hypothesis
**Hypothesis 1**

`Null Hypothesis (Ho):` There is no significant difference in churn rates between customers with higher and lower monthly charge.

`Alternative Hypothesis (Ha):` There is a significant difference in churn rates between customers with higher and lower monthly charge.


### 1.6 Business Questions
1. What is the average tenure of customers who churned compared to those who stayed?
2. Do customers with partners or dependents have a lower churn rate?
3. Is there a correlation between the contract term (Contract) and customer churn?
4. What are the common payment methods (Payment Method) among customers who churned?
5. How does the availability of tech-related services (e.g., OnlineSecurity, TechSupport) impact churn rates?
6. What percentage of customers who churned had streaming services (StreamingTV, StreamingMovies)?
7. How does the total amount charged to customers (TotalCharges) correlate with churn behavior?
## 2.0 Data Understanding

### 2.1 Inspecting the dataset in depth.

A. Data Quality Assessment(info, duplicates, null values, describe etc.)

B. Univariate Analysis: Explore, analyze and visualize key variables independently of others

C. Bivariate Analysis: Explore, analyze and visualize the relationship variables pairs of different dimensions

D. Multivariate Analysis: Explore, analyze and visualize the relationship among variables

E. Answer Analytical Questions

F. Test Hypothesis:

G. Provide insights:

In [1]:
#pip install openpyxl

In [2]:
#import libraries
import pandas as pd
import sqlalchemy as sa 
from dotenv import load_dotenv, dotenv_values
import pyodbc
import warnings

In [3]:
warnings.filterwarnings("ignore")

In [4]:
# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values('.env_telco_churn_db')

# Get the values for the credentials you set in the '.env' file
server = environment_variables.get("SERVER")
database = environment_variables.get("DATABASE")
username = environment_variables.get("USERNAME")
password = environment_variables.get("PASSWORD")

In [5]:
username

'LP2_project'

In [6]:
# Create a connection string
connection_string = f"DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password};MARS_Connection=yes;MinProtocolVersion=TLSv1.2;"

In [7]:
# Use the connect method of the pyodbc library and pass in the connection string.
# This will connect to the server and might take a few seconds to be complete. 
# Check your internet connection if it takes more time than necessary

connection = pyodbc.connect(connection_string)

In [8]:
# Now the sql query to get the data is what what you see below. 
# Note that you will not have permissions to insert delete or update this database table. 

query = "SELECT * FROM LP2_Telco_churn_first_3000"

data = pd.read_sql(query, connection)

In [9]:
data.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG  Female          False     True       False       1   
1  5575-GNVDE    Male          False    False       False      34   
2  3668-QPYBK    Male          False    False       False       2   
3  7795-CFOCW    Male          False    False       False      45   
4  9237-HQITU  Female          False    False       False       2   

   PhoneService MultipleLines InternetService OnlineSecurity  ...  \
0         False          None             DSL          False  ...   
1          True         False             DSL           True  ...   
2          True         False             DSL           True  ...   
3         False          None             DSL           True  ...   
4          True         False     Fiber optic          False  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0            False       False       False           False  Month-to-month   
1             True       False       False           False        One year   
2            False       False       False           False  Month-to-month   
3             True        True       False           False        One year   
4            False       False       False           False  Month-to-month   

  PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0             True           Electronic check      29.850000     29.850000   
1            False               Mailed check      56.950001   1889.500000   
2             True               Mailed check      53.849998    108.150002   
3            False  Bank transfer (automatic)      42.299999   1840.750000   
4             True           Electronic check      70.699997    151.649994   

   Churn  
0  False  
1  False  
2   True  
3  False  
4   True  

[5 rows x 21 columns]

In [10]:
data.isna().sum()

customerID            0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines       269
InternetService       0
OnlineSecurity      651
OnlineBackup        651
DeviceProtection    651
TechSupport         651
StreamingTV         651
StreamingMovies     651
Contract              0
PaperlessBilling      0
PaymentMethod         0
MonthlyCharges        0
TotalCharges          5
Churn                 1
dtype: int64

In [11]:
#load second dataset
data2 = pd.read_csv("data/LP2_Telco-churn-second-2000.csv")
data2.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  5600-PDUJF    Male              0      No         No       6          Yes   
1  8292-TYSPY    Male              0      No         No      19          Yes   
2  0567-XRHCU  Female              0     Yes        Yes      69           No   
3  1867-BDVFH    Male              0     Yes        Yes      11          Yes   
4  2067-QYTCF  Female              0     Yes         No      64          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0                No             DSL             No  ...               No   
1                No             DSL             No  ...              Yes   
2  No phone service             DSL            Yes  ...              Yes   
3               Yes     Fiber optic             No  ...               No   
4               Yes     Fiber optic             No  ...              Yes   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes          No              No  Month-to-month              Yes   
1         Yes          No              No  Month-to-month              Yes   
2          No          No             Yes        Two year              Yes   
3          No          No              No  Month-to-month              Yes   
4         Yes         Yes             Yes  Month-to-month              Yes   

             PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Credit card (automatic)          49.50         312.7    No  
1  Credit card (automatic)          55.00        1046.5   Yes  
2  Credit card (automatic)          43.95        2960.1    No  
3         Electronic check          74.35         834.2   Yes  
4         Electronic check         111.15        6953.4    No  

[5 rows x 21 columns]

In [12]:
#compare columns
print(data.columns)
print(data2.columns)

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')
Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')


In [13]:
#load test dataset
test_data = pd.read_excel("data/Telco-churn-last-2000.xlsx")
test_data.head()

customerID gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7613-LLQFO   Male              0      No         No      12          Yes   
1  4568-TTZRT   Male              0      No         No       9          Yes   
2  9513-DXHDA   Male              0      No         No      27          Yes   
3  2640-PMGFL   Male              0      No        Yes      27          Yes   
4  3801-HMYNL   Male              0     Yes        Yes       1          Yes   

  MultipleLines InternetService       OnlineSecurity         OnlineBackup  \
0           Yes     Fiber optic                   No                   No   
1            No              No  No internet service  No internet service   
2            No             DSL                  Yes                   No   
3           Yes     Fiber optic                   No                   No   
4            No     Fiber optic                   No                   No   

      DeviceProtection          TechSupport          StreamingTV  \
0                   No                   No                  Yes   
1  No internet service  No internet service  No internet service   
2                  Yes                  Yes                  Yes   
3                   No                  Yes                   No   
4                   No                   No                  Yes   

       StreamingMovies        Contract PaperlessBilling     PaymentMethod  \
0                   No  Month-to-month              Yes  Electronic check   
1  No internet service  Month-to-month               No      Mailed check   
2                  Yes        One year               No  Electronic check   
3                   No  Month-to-month              Yes  Electronic check   
4                  Yes  Month-to-month               No      Mailed check   

   MonthlyCharges TotalCharges  
0           84.45      1059.55  
1           20.40        181.8  
2           81.70      2212.55  
3           79.50      2180.55  
4           89.15        89.15

In [14]:
#join data and data1 DataFrame

train_data = pd.concat([data,data2], ignore_index=True)
train_data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0    True      False       1        False   
1  5575-GNVDE    Male              0   False      False      34         True   
2  3668-QPYBK    Male              0   False      False       2         True   
3  7795-CFOCW    Male              0   False      False      45        False   
4  9237-HQITU  Female              0   False      False       2         True   

  MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0          None             DSL          False  ...            False   
1         False             DSL           True  ...             True   
2         False             DSL           True  ...            False   
3          None             DSL           True  ...             True   
4         False     Fiber optic          False  ...            False   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0       False       False           False  Month-to-month             True   
1       False       False           False        One year            False   
2       False       False           False  Month-to-month             True   
3        True       False           False        One year            False   
4       False       False           False  Month-to-month             True   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check      29.850000         29.85  False  
1               Mailed check      56.950001        1889.5  False  
2               Mailed check      53.849998    108.150002   True  
3  Bank transfer (automatic)      42.299999       1840.75  False  
4           Electronic check      70.699997    151.649994   True  

[5 rows x 21 columns]

In [15]:
train_data.shape

(5043, 21)

In [16]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        5043 non-null   object 
 1   gender            5043 non-null   object 
 2   SeniorCitizen     5043 non-null   int64  
 3   Partner           5043 non-null   object 
 4   Dependents        5043 non-null   object 
 5   tenure            5043 non-null   int64  
 6   PhoneService      5043 non-null   object 
 7   MultipleLines     4774 non-null   object 
 8   InternetService   5043 non-null   object 
 9   OnlineSecurity    4392 non-null   object 
 10  OnlineBackup      4392 non-null   object 
 11  DeviceProtection  4392 non-null   object 
 12  TechSupport       4392 non-null   object 
 13  StreamingTV       4392 non-null   object 
 14  StreamingMovies   4392 non-null   object 
 15  Contract          5043 non-null   object 
 16  PaperlessBilling  5043 non-null   object 


The columns present in the dataset are:
 
- **Gender**: Whether the customer is a male or a female
 
- **SeniorCitizen**: Whether a customer is a senior citizen or not
 
- **Partner**: Whether the customer has a partner or not (True, False)
 
- **Dependents**: Whether the customer has dependents or not ((True, False)
 
- **Tenure**: Number of months the customer has stayed with the company
 
- **Phone Service**: Whether the customer has a phone service or not ((True, False
 
- **MultipleLines**: Whether the customer has multiple lines or not
 
- **InternetService**: Customer's internet service provider (DSL, Fiber Optic, No)
 
- **OnlineSecurity**: Whether the customer has online security or not ((True, False, No Internet)
 
- **OnlineBackup**: Whether the customer has online backup or not ((True, False, No Internet)
 
- **DeviceProtection**: Whether the customer has device protection or not ((True, False, No internet service)
 
- **TechSupport**: Whether the customer has tech support or not ((True, False, No internet)
 
- **StreamingTV**: Whether the customer has streaming TV or not ((True, False, No internet service)
 
- **StreamingMovies**: Whether the customer has streaming movies or not ((True, False, No Internet service)
 
- **Contract**: The contract term of the customer (Month-to-Month, One year, Two year)
 
- **PaperlessBilling**: Whether the customer has paperless billing or not ((True, False)
 
- **Payment Method**: The customer's payment method (Electronic check, mailed check, Bank transfer(automatic), Credit card(automatic))
 
- **MonthlyCharges**: The amount charged to the customer monthly
 
- **TotalCharges**: The total amount charged to the customer
 
- **Churn**: Whether the customer churned or not ((True, False)

In [17]:
#convert TotalCharges column to float for both train and test dataset
train_data["TotalCharges"] = pd.to_numeric(train_data["TotalCharges"], errors="coerce")
test_data["TotalCharges"] = pd.to_numeric(test_data["TotalCharges"], errors="coerce")

print(train_data["TotalCharges"].dtype)
print(test_data["TotalCharges"].dtype)

float64
float64


In [18]:
#chechk for missing values
train_data.isna().sum()

customerID            0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines       269
InternetService       0
OnlineSecurity      651
OnlineBackup        651
DeviceProtection    651
TechSupport         651
StreamingTV         651
StreamingMovies     651
Contract              0
PaperlessBilling      0
PaymentMethod         0
MonthlyCharges        0
TotalCharges          8
Churn                 1
dtype: int64

### Key Observations:
 
These columns contain **missing data**
    `MultipleLines`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`,  `StreamingMovies`, `TotalCharges`, '`Churn`
 
 `MultipleLines` has 269 missing values
 `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`,  `StreamingMovies` has 651 missing values.
 
 `TotalCharges`     has 8 missing values
 
 `Churn`            has 1 missing value

In [19]:
def summary(df):
    unique_summary = []
    #iterate through each column
    for col in df.columns:
        #get unique values of the column
        unique_values = df[col].unique()
        #get the count of the unique values
        unique_count = len(unique_values)
        #append evrything to the empty list
        unique_summary.append({
            "Column": col,
            "Unique value count": unique_count,
            "Unique values": unique_values
        })
    summary_df = pd.DataFrame(unique_summary)

    return summary_df


In [20]:
summary(train_data)

Column  Unique value count  \
0         customerID                5043   
1             gender                   2   
2      SeniorCitizen                   2   
3            Partner                   4   
4         Dependents                   4   
5             tenure                  73   
6       PhoneService                   4   
7      MultipleLines                   6   
8    InternetService                   3   
9     OnlineSecurity                   6   
10      OnlineBackup                   6   
11  DeviceProtection                   6   
12       TechSupport                   6   
13       StreamingTV                   6   
14   StreamingMovies                   6   
15          Contract                   3   
16  PaperlessBilling                   4   
17     PaymentMethod                   4   
18    MonthlyCharges                2069   
19      TotalCharges                4864   
20             Churn                   5   

                                        Unique values  
0   [7590-VHVEG, 5575-GNVDE, 3668-QPYBK, 7795-CFOC...  
1                                      [Female, Male]  
2                                              [0, 1]  
3                              [True, False, No, Yes]  
4                              [False, True, No, Yes]  
5   [1, 34, 2, 45, 8, 22, 10, 28, 62, 13, 16, 58, ...  
6                              [False, True, Yes, No]  
7      [None, False, True, No, No phone service, Yes]  
8                              [DSL, Fiber optic, No]  
9   [False, True, None, No, Yes, No internet service]  
10  [True, False, None, No, Yes, No internet service]  
11  [False, True, None, No, Yes, No internet service]  
12  [False, True, None, Yes, No, No internet service]  
13  [False, True, None, No, Yes, No internet service]  
14  [False, True, None, No, Yes, No internet service]  
15               [Month-to-month, One year, Two year]  
16                             [True, False, Yes, No]  
17  [Electronic check, Mailed check, Bank transfer...  
18  [29.850000381469727, 56.95000076293945, 53.849...  
19  [29.850000381469727, 1889.5, 108.1500015258789...  
20                       [False, True, None, No, Yes]

#### Recommendations:
- Drop the `CustomerID` column.
- The count of unique values tends to be highest in numerical columns.
- Standardize the unique values of categorical columns to allow consistency by casting to "Yes" or "No" where implied. Therefore, False, 0, "No phone service" and "No internet service" become "No", while True & 1 becomes "Yes". (13 columns)
- "None" values are missing or null values so replace NULL/None with pd.NA  
- Consider visualization techniques such as histograms, bar plots, or box plots to gain deeper insights into the distribution of categorical and numerical variables.

In [21]:
#drop the column customer ID
train_data.drop(columns="customerID", axis=1, inplace=True)

In [22]:
#display all columns with objest data type
categorical = train_data.select_dtypes(include="object").columns.tolist()
categorical

['gender',
 'Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Churn']

In [23]:
#convert colums with objest data type to string
train_data[categorical] = train_data[categorical].astype(str)

In [24]:
yep = {
    "No": ["False", "0", "No phone service", "No internet service"],
    "Yes": ["True", "1"]
}
 
# Function to make corrections on a DataFrame
def correct_columns(train_data, categorical, yep):
   
    for column_name in categorical:
        for key, values in yep.items():
            # Replace values in the DataFrame for each column
            train_data[column_name] = train_data[column_name].replace(values, key)
    return train_data
 
# Apply corrections to the DataFrame
train_data = correct_columns(train_data, categorical, yep)

In [25]:
summary(train_data)

Column  Unique value count  \
0             gender                   2   
1      SeniorCitizen                   2   
2            Partner                   2   
3         Dependents                   2   
4             tenure                  73   
5       PhoneService                   2   
6      MultipleLines                   3   
7    InternetService                   3   
8     OnlineSecurity                   3   
9       OnlineBackup                   3   
10  DeviceProtection                   3   
11       TechSupport                   3   
12       StreamingTV                   3   
13   StreamingMovies                   3   
14          Contract                   3   
15  PaperlessBilling                   2   
16     PaymentMethod                   4   
17    MonthlyCharges                2069   
18      TotalCharges                4864   
19             Churn                   3   

                                        Unique values  
0                                      [Female, Male]  
1                                              [0, 1]  
2                                           [Yes, No]  
3                                           [No, Yes]  
4   [1, 34, 2, 45, 8, 22, 10, 28, 62, 13, 16, 58, ...  
5                                           [No, Yes]  
6                                     [None, No, Yes]  
7                              [DSL, Fiber optic, No]  
8                                     [No, Yes, None]  
9                                     [Yes, No, None]  
10                                    [No, Yes, None]  
11                                    [No, Yes, None]  
12                                    [No, Yes, None]  
13                                    [No, Yes, None]  
14               [Month-to-month, One year, Two year]  
15                                          [Yes, No]  
16  [Electronic check, Mailed check, Bank transfer...  
17  [29.850000381469727, 56.95000076293945, 53.849...  
18  [29.850000381469727, 1889.5, 108.1500015258789...  
19                                    [No, Yes, None]

In [26]:
#get columns with number data type
numerical = train_data.select_dtypes(include=["number"])
numerical

SeniorCitizen  tenure  MonthlyCharges  TotalCharges
0                 0       1       29.850000     29.850000
1                 0      34       56.950001   1889.500000
2                 0       2       53.849998    108.150002
3                 0      45       42.299999   1840.750000
4                 0       2       70.699997    151.649994
...             ...     ...             ...           ...
5038              0      24       84.800000   1990.500000
5039              0      72      103.200000   7362.900000
5040              0      11       29.600000    346.450000
5041              1       4       74.400000    306.600000
5042              0      66      105.650000   6844.500000

[5043 rows x 4 columns]

In [27]:
#get stats on numerical columns
numerical.describe().T

count         mean          std        min         25%  \
SeniorCitizen   5043.0     0.162403     0.368857   0.000000    0.000000   
tenure          5043.0    32.576641    24.529807   0.000000    9.000000   
MonthlyCharges  5043.0    65.090214    30.068133  18.400000   35.775000   
TotalCharges    5035.0  2302.062861  2269.483576  18.799999  417.675003   

                        50%          75%          max  
SeniorCitizen      0.000000     0.000000     1.000000  
tenure            29.000000    56.000000    72.000000  
MonthlyCharges    70.550003    90.050003   118.650002  
TotalCharges    1401.150024  3862.000000  8670.100000

In [28]:
#get stats on categorical columns
train_data.describe(include="object").T

count unique               top  freq
gender            5043      2              Male  2559
Partner           5043      2                No  2585
Dependents        5043      2                No  3482
PhoneService      5043      2               Yes  4554
MultipleLines     5043      3                No  2644
InternetService   5043      3       Fiber optic  2248
OnlineSecurity    5043      3                No  2936
OnlineBackup      5043      3                No  2660
DeviceProtection  5043      3                No  2648
TechSupport       5043      3                No  2901
StreamingTV       5043      3                No  2447
StreamingMovies   5043      3                No  2410
Contract          5043      3    Month-to-month  2744
PaperlessBilling  5043      2               Yes  2996
PaymentMethod     5043      4  Electronic check  1700
Churn             5043      3                No  3706

In [29]:
numerical.isna().sum()

SeniorCitizen     0
tenure            0
MonthlyCharges    0
TotalCharges      8
dtype: int64

In [ ]:
#display columns with null values in the column TotalCharge
train_data[train_data["TotalCharges"].isna()]

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
488   Female              0     Yes        Yes       0           No   
753     Male              0      No        Yes       0          Yes   
936   Female              0     Yes        Yes       0          Yes   
1082    Male              0     Yes        Yes       0          Yes   
1340  Female              0     Yes        Yes       0           No   
3218    Male              0     Yes        Yes       0          Yes   
4670  Female              0     Yes        Yes       0          Yes   
4754    Male              0      No        Yes       0          Yes   

     MultipleLines InternetService OnlineSecurity OnlineBackup  \
488           None             DSL            Yes           No   
753             No              No           None         None   
936             No             DSL            Yes          Yes   
1082           Yes              No           None         None   
1340          None             DSL            Yes          Yes   
3218            No              No             No           No   
4670           Yes             DSL             No          Yes   
4754           Yes             DSL            Yes          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies  Contract  \
488               Yes         Yes         Yes              No  Two year   
753              None        None        None            None  Two year   
936               Yes          No         Yes             Yes  Two year   
1082             None        None        None            None  Two year   
1340              Yes         Yes         Yes              No  Two year   
3218               No          No          No              No  One year   
4670              Yes         Yes         Yes              No  Two year   
4754               No         Yes          No              No  Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
488               Yes  Bank transfer (automatic)       52.549999   
753                No               Mailed check       20.250000   
936                No               Mailed check       80.849998   
1082               No               Mailed check       25.750000   
1340               No    Credit card (automatic)       56.049999   
3218              Yes               Mailed check       19.700000   
4670               No               Mailed check       73.350000   
4754              Yes  Bank transfer (automatic)       61.900000   

      TotalCharges Churn  
488            NaN    No  
753            NaN    No  
936            NaN    No  
1082           NaN    No  
1340           NaN    No  
3218           NaN    No  
4670           NaN    No  
4754           NaN    No

In [30]:
import numpy as np

### Imputation of missing values

In [35]:
#fill null values in TotalCharges with median (median is not acceptable to outliers)
numerical["TotalCharges"].fillna(numerical["TotalCharges"].median(), inplace=True)

In [36]:
numerical.isna().sum()

SeniorCitizen     0
tenure            0
MonthlyCharges    0
TotalCharges      0
dtype: int64

In [46]:
summary(train_data[["PaymentMethod"]])

Column  Unique value count  \
0  PaymentMethod                   4   

                                       Unique values  
0  [Electronic check, Mailed check, Bank transfer...